src/utils/schemaTree.js

```diff
@@ -0,0 +1,312 @@
+import { clone, assocPath } from 'ramda';
+import { COMBINATION_TYPES, COMPLEX_TYPES } from './constants';
+
+/**
+ * Generate a tree that illustrates the recursive structure of a schema.
+ * A single node within the tree defines the following structure:
+ * {
+ *   schema: ..       // the schema the node is representing
+ *   children: [..]   // node objects of the same structure (for subschemas)
+ *   path: [..]       // the path from root to node, array of child indexes in
+ *                       sequential order (for $refs and indentation level)
+ *                       ex. [1, 0] = node at root.children[1].children[0]
+ *                                    also has depth of 1 for indent size.
+ * }
+ *
+ * If node is a ref node, it may have a difference structure:
+ * {
+ *   defaultNode: ..  // the default node structure when a $ref is shrinked
+ *                       (a node object as illustrated above)
+ *   expandedNode: .. // the expanded node strucutre when a $ref is expanded
+ *                       (a node object as illustrated above)
+ *   isExpanded: ..   // whether the node is expanded or not

```

Comment:
refNodes` have a bit of a different structure from a normal `treeNode` in order to cache both the expanded form and the default collapsed form. 

- The `defaultNode` holds the default form of a node and has the same structure of a normal treeNode. The `expandedNode` holds the expanded form of a node and also has the same structure of a normal treeNode. So basically, `refNode` stores two `treeNode`s just in different forms so that it is easy for the ref rows to switch between collapsed and expanded forms.
- Also, the reason why the actual node objects are stored in the `refNode`, instead of just storing the schemas of each, is because `NormalLeftRow` and `NormalRightRow` create rows based on a node structures.